skglm/solvers/group_bcd_solver.py

```diff
@@ -0,0 +1,100 @@
+import numpy as np
+from numba import njit
+
+
+def grp_bcd_solver(X, y, datafit, penalty,
+                   max_iter=1000, tol=1e-7, verbose=False):
+    """Run a group BCD solver.
+
+    Parameters
+    ----------
+    X : array, shape (n_samples, n_features)
+        Design matrix.
+
+    y : array, shape (n_samples,)
+        Target vector.
+
+    datafit : QuadraticGroup
+        DataFit object.
+
+    penalty : WeightedGroupL1
+        Penalty object.
+
+    max_iter : int, default 1000
+        Maximum number of iterations.
+
+    tol : float, default 1e-7
+        Tolerance for convergence.
+
+    verbose : bool, default False
+        Log or not the objective at each iteration.
+
+    Returns
+    -------
+    w : array, shape (n_features,)
+        Vector that minimizes ``datafit() + penalty()``.
+    """
+    n_samples, n_features = X.shape
+    n_groups = len(penalty.grp_partition) - 1
+
+    # init
+    w = np.zeros(n_features)
+    Xw = np.zeros(n_samples)  # X @ w in general
+    datafit.initialize(X, y)
+    all_groups = np.arange(n_groups)
+
+    for k in range(max_iter):
+        prev_p_obj = datafit.value(y, w, Xw) + penalty.value(w)
+        _bcd_epoch(X, y, w, Xw, datafit, penalty, all_groups)
+        current_obj = datafit.value(y, w, Xw) + penalty.value(w)
+
+        if verbose:
+            print(f"Iteration {k}: {current_obj}")
+
+        if np.abs(current_obj - prev_p_obj) < tol:  # naive stopping criterion
+            print("Early exit")
+            break
+
+    return w
+
+
+@njit
+def _bcd_epoch(X, y, w, Xw, datafit, penalty, ws):
+    """Perform a single BCD epoch on groups in ws.
+
+    Parameters
```

Comment:
IMO, we don't need to document private functions. It induces a higher maintenance cost